metricflow/query/query_parser.py

```diff
@@ -376,8 +377,26 @@ def _parse_and_validate_query(
                 metric_references=metric_references, time_dimension_specs=where_time_specs
             )
 
+        metric_specs: List[MetricSpec] = []
+        for metric_reference in base_metric_references:
```

Comment:
Hmmm. It was a little hard to wrap my head around this.

The reason we do base_metric_references is because we only evaluate constraints on metrics at this point. The input_metrics to a DERIVED metric have the constraints pushed into the spec inside MetricSemantics.

It might be worth it to consolidate this further - right now we do top level spec construction in the query parser and upstream dependent spec construction inside MetricSemantics. But I also think that's likely opening up a can of worms, since MetricSemantics is in need of some restructuring itself.

So, I've convinced myself we should leave this here. But let's add a comment somewhere in this vicinity for why the base_metric_references are the things we care about here.

Also, I suggest initializing the local variable right above this block, it's not as important as putting it on line 233 makes it seem. We might even want to factor this all out into a helper method, then use the docstring to explain everything that's going on. I think I like that best.